### ISSUE TITLE
Question about `MultiIndex`

### ISSUE THREAD

**Issue by durkmurder** [CosmWasm/cw-plus#466]
Hey, I am struggling to make `MultiIndex` work the way I need, maybe it's not possible at all at this stage. Hope to get some advice.

Basically I have a primary key which is a tuple `(Addr, String)` and additional index `collection` which is a `String`. I need to be able to query by secondary index(by collection) and do filtering and pagination by primary key.
I can do this with two maps: `Primary: Map<(Addr, String), Value>` + `SecondaryIndex: Map<(String, (Addr, String)), Empty>` and then I can do something like this:

```
SecondaryIndex
.prefix(collection)
.keys(...)
.filter(|(_, (addr, s)| addr == filter_addr)
.take(1)
.map(|(_, (addr, s) Primary.prefix(addr).range(Bound(s), None).map(// do something with result)
```
This is probably doable but not very efficient and hard to maintain I would like to do something:

```
get_indexed_map()
.idx.collection.prefix(collection)
// here I would like to prefix primary key with filter_addr and then call range to do pagination
.range(
deps.storage,
Bound(s),
None,
Order::Ascending,
```

```
pub struct ExtraIndexes<'a> {
    // pk goes to second tuple element
    pub collection: MultiIndex<'a, (String, Vec<u8>), Value>,
}

get_indexed_map() -> IndexedMap<'a, (Addr, String), Value, ExtraIndexes<'a>>
```

Any advice is appreciated.

**Comment by maurolacy** [comment]
This should work. I recommend you to follow the docs at https://docs.cosmwasm.com/tutorials/storage/indexes/

If you have specific questions / code snippets, we could help you better.

**Comment by durkmurder** [comment]
I have checked those docs, but it doesn't give me an answer. 
I will give more concrete example:
### Sample:
```
pub struct OwnedTokensIndexes<'a> {
    // pk goes to second tuple element
    pub collection: MultiIndex<'a, (String, Vec<u8>), OwnedTokensInfo>,
}

impl<'a> IndexList<OwnedTokensInfo> for OwnedTokensIndexes<'a> {
    fn get_indexes(&'_ self) -> Box<dyn Iterator<Item = &'_ dyn Index<OwnedTokensInfo>> + '_> {
        let v: Vec<&dyn Index<OwnedTokensInfo>> = vec![&self.collection];
        Box::new(v.into_iter())
    }
}

/// Mapping that represents (Owner, TokenId) -> OwnedTokensInfo
/// It used to quickly query tokens by owner without counting editions.
pub fn owned_tokens<'a>() -> IndexedMap<'a, (Addr, String), OwnedTokensInfo, OwnedTokensIndexes<'a>>
{
    let indexes = OwnedTokensIndexes {
        collection: MultiIndex::new(
            |d: &OwnedTokensInfo, k: Vec<u8>| (d.collection.clone(), k),
            "owned_tokens",
            "owned_tokens__collection",
        ),
    };
    IndexedMap::new("owned_tokens", indexes)
}
```

So here my value is indexed by `(Owner, TokenId)` and `Collection`. I would like to benefit from that:
I can write:

```
let start_after = Some("some_token_id");
owned_tokens().prefix(owner_addr).range(
            deps.storage,
            start_after.map(Bound::exclusive),
            None,
            Order::Ascending,
),
```
This will work as expected, I will first filter by address using `prefix(owner_addr)` and then use range to paginate over second element of tuple.
How should I approach this if I am looking up index first?
```
owned_tokens().idx.collection.prefix(collection).range(
            deps.storage,
            None, // ? what should go here? is is range over (Owner, TokenId) ? How to construct one ?
            None,
            Order::Ascending,
),
```
How do I prefix and then use range to get the same functionality as I was getting without using multi index.


### Bonus Question:
How to parse Vec<u8> as primary key to the actual type? I've noticed when tuples are used it contains extra bytes in the beginning.
```
owned_tokens().idx.collection.prefix(collection).keys(
            deps.storage,
            None,
            None,
            Order::Ascending,
).map (| v | // v is Vec<u8> how to parse it?)
```

**Comment by maurolacy** [comment]
> None, // ? what should go here? is is range over (Owner, TokenId) ? How to construct one ?

The primary key (that is, the suffix of the full key, after you prefixed it), in raw format, because the 2nd element of the index is a `Vec<u8>`.

To construct one, and assuming this is a `start_after` pagination param that comes as a `String`, you can use `start_after.into_bytes()`.

If it comes as an `Addr` (because you validate it or so), you can do `start_after.to_string().into_bytes()` or similar.

If this is a pagination parameter, it's usually enough, when iterating in the client, to use the last received value (while it's non-null) and send it as the `start_after` param. Then you use as the start of `range` with an exclusive bound.

**Comment by maurolacy** [comment]
> ).map (| v | // v is Vec<u8> how to parse it?)

Use `from_utf8` to re-convert to an owned `String`, and return it.

**Comment by maurolacy** [comment]
There are examples of this in in the `cw-plus` contracts.

We're are also working on a version of `range` with full key deserialization. So that you can receive an `Addr` there, instead of a `Vec<u8>`. This is already implemented for `Map`, and we plan to extend support to indexes soon.

So, stay tuned.

**Comment by durkmurder** [comment]
> > ).map (| v | // v is Vec how to parse it?)
> 
> Use `from_utf8` to re-convert to an owned `String`, and return it.

but it's not a string, my primary key is `(Addr, String)` so `Vec<u8>` will have something like `[0, 1, bytes for addr, bytes for string]` at least 2 extra bytes at the beginning.

**Comment by durkmurder** [comment]
> > None, // ? what should go here? is is range over (Owner, TokenId) ? How to construct one ?
> 
> The primary key (that is, the suffix of the full key, after you prefixed it). In raw format, because the 2nd element of the index is a `Vec<u8>`.
> 
> To construct one, and assuming this is a `start_after` pagination param that comes as a `String`, you can use `start_after.into_bytes()`.
> 
> If it comes as an `Addr` (because you validated it or so), you can do `start_after.to_string().into_bytes()` or similar.
> 
> If this is a pagination parameter, it's usually enough, when iterating in the client, to use the last received value (while it's non-null) and send it as the `start_after` param. Then you use it as the start of `range` with an exclusive bound.

`owned_tokens().idx.collection.prefix(collection)` at this point I can call `keys` or `range` and both can accept `start` and `end`. 
I assume that `start` should be serialized primary key. Since my primary key is a tuple I would like to `prefix` using first element(addr) and `range` over second element(token_id), is that possible? In other words I need filtering by address and pagination by token_id

**Comment by maurolacy** [comment]
> > > None, // ? what should go here? is is range over (Owner, TokenId) ? How to construct one ?
> > 
> > 
> > The primary key (that is, the suffix of the full key, after you prefixed it). In raw format, because the 2nd element of the index is a `Vec<u8>`.
> > To construct one, and assuming this is a `start_after` pagination param that comes as a `String`, you can use `start_after.into_bytes()`.
> > If it comes as an `Addr` (because you validated it or so), you can do `start_after.to_string().into_bytes()` or similar.
> > If this is a pagination parameter, it's usually enough, when iterating in the client, to use the last received value (while it's non-null) and send it as the `start_after` param. Then you use it as the start of `range` with an exclusive bound.
> 
> `owned_tokens().idx.collection.prefix(collection)` at this point I can call `keys` or `range` and both can accept `start` and `end`. I assume that `start` should be serialized primary key. Since my primary key is a tuple I would like to `prefix` using first element(addr) and `range` over second element(token_id), is that possible? In other words I need filtering by address and pagination by token_id

No. `range` here is over the remaining elements of the index, namely, the full primary key, which is a tuple. You _can_ `range` over it, but will have to filter manually afterwards.

A way to build the bound is `Bound::exclusive((owner, token_id).joined_key())`. There are other ways, but this is the more concise.

**Comment by durkmurder** [comment]
Thanks a lot for responses, now it makes sense. I had suspicions that it's the case but needed to verify it. Any chance that you could introduce `prefix` after querying the index? Would really help a lot + I am really waiting for deserialization updates(I've seen PR in progress).

**Comment by maurolacy** [comment]
In any case, take a look at the code above. It will work, and properly support pagination / continuation.

I mentioned some stuff about fixed vs. variable length addresses, that I now deleted, because but that's only an issue in case you want to continue iterating _past_ `owner`, and have variable length strings.

**Comment by durkmurder** [comment]
> In any case, take a look at the code above. It will work, and properly support pagination / continuation.
> 
> I mentioned some stuff about fixed vs. variable length addresses, that I now deleted, because that's only an issue in case you want to continue iterating _past_ `owner`, and have variable length strings.

I am trying to utilize `U64Key` after your suggestions, is there a function to deserialize Vec<u8> into `U64Key`? I need it to access key in `range().map()`

**Comment by maurolacy** [comment]
Just use `let id = u64::from_be_bytes(value.as_slice().try_into()...)` and you'll recover the number. Then `U64Key::new(id)`, and you have it.

**Comment by durkmurder** [comment]
> Just use `let id = u64::from_be_bytes(value.as_slice().try_into()...)` and you'll recover the number. Then `U64Key::new(id)`, and you have it.

Thanks.
I am doing the same with `Endian::from_be_bytes`, seems does the trick.
